text/3735-field-projections.md

```diff
@@ -0,0 +1,1357 @@
+        let mut buf: Pin<Guard<'_, Buffer, MutexBackend>> = self.buf.lock();
+        // This method allocates, so it must be fallible.
+        // `buf.as_mut()->buf` again uses the field projection for `Pin` to yield a `&mut Vec<u8>`.
+        buf.as_mut()->buf.extend_from_slice(data)
+    }
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Rust Book Chapter: Field Projections
+
+When programming in Rust, one often has the need to access only a single field of a struct. In the
+usual cases of `&T` or `&mut T`, this is simple. Just use dot syntax and you can create a reference
+to a field of the struct `&t.field`.
+
+However, when one has a different type that "contains" or "points" at a `T`, one has to reach for
+*field projections* via the *field projection operator* `->`. In this chapter, we will learn what
+field projections are and how to use them for the most common types from the standard library. For
+example for pointer-like types and [pin projections].
+
+### Simple Uses of Field Projections
+
+Let's say we have a big struct that doesn't fit onto the stack:
+
+```rust
+struct Data {
+    flags: u32,
+    buf: [u8; 1024 * 1024],
+}
+```
+
+We would like to initialize the bytes in `buf` to `0xff` and `flags` should be `0x0f`. We start with
+a new function returning memory on the heap:
+
+```rust
+impl Data {
+    fn new() -> Box<Data> {
+        let mut data = Box::new_uninit();
+        {
+            let data: &mut MaybeUninit<Data> = &mut *data;
+```
+
+Now we can use field projection to turn `&mut MaybeUninit<Data>` into `&mut MaybeUninit<u32>` that
+points to the `flags` field:
+
+```rust
+            let flags: &mut MaybeUninit<u32> = data->flags;
+            flags.write(0x0f);
+```
+
+And to initialize `buf`, we can do the same:
+
+```rust
+            let buf: &mut MaybeUninit<[u8]> = data->buf;
+            let buf: &mut [MaybeUninit<u8>] = MaybeUninit::slice_as_bytes_mut(buf);
+            MaybeUninit::fill(buf, 0xff);
+        }
+```
+
+Now we only need to unsafely assert that we initialized everything.
+
+```rust
+        unsafe { data.assume_init() }
+    }
+}
+```
+
+A more general explanation of field projection is that it is an operation that turns a generic
+container type `C<T>` containing a struct `T` into a container `C<F>` where `F` is a field of the
+struct `T`.
+
+#### Raw Pointers
+
+Similarly to `&mut MaybeUninit<T>`, raw pointers also support projections. Given a raw pointer
+`ptr: *mut Data`, one can use field projection to obtain a pointer to a field:
+`ptr->flags: *mut u32`. Essentially `ptr->field` is a shorthand for `&raw mut (*ptr).field` (for
+`*const` the same is true except for the `mut`). However, there is a small difference between the
+two: the latter has to be `unsafe`, since `*ptr` requires that `ptr` be dereferencable. But field
+projection is a safe operation and thus it uses [`wrapping_add`] under the hood. This is less
```

Comment:
I think it might be better for raw-pointer(-like) field projection to be unsafe (like the "desugared" version), and require the at least the specific offset to be in-bounds (if not the whole container).

In particular, `NonNull` having field projection be safe would mean that it would have to panic (or something) if the projection would result in null.

It seems like none of the motivating examples would benefit from allowing `ptr->field` where `field` is not inbounds? Maybe there are usecases for this but I haven't thought of any where it wouldn't be clearer to explicitly use `wrapping_add` IMO.

Reply:
> I think it might be better for raw-pointer(-like) field projection to be unsafe (like the "desugared" version), and require the at least the specific offset to be in-bounds (if not the whole container).
> 
> In particular, `NonNull` having field projection be safe would mean that it would have to panic (or something) if the projection would result in null.

Note that `wrapping_add` is itself a safe function. So the implementation of `NonNull::project` only contains safe code, making UB impossible. Also, there is no need for a panic, since a projection can never result in null.

> It seems like none of the motivating examples would benefit from allowing `ptr->field` where `field` is not inbounds? Maybe there are usecases for this but I haven't thought of any where it wouldn't be clearer to explicitly use `wrapping_add` IMO.

The reason for why I used `wrapping_add` is that the following code is UB when the `->` operation uses `add`. Making the operation `unsafe` for raw pointers is probably impossible, since it is backed by the `Project` trait and the `project` function can either be `unsafe` or safe.
Maybe we can split it into two different mutually exclusive traits, but I think it would be better both for consistency and for ergonomic reasons to keep field projections safe.

Reply:
> Note that `wrapping_add` is itself a safe function. So the implementation of `NonNull::project` only contains safe code, making UB impossible. Also, there is no need for a panic, since a projection can never result in null.

`NonNull::wrapping_add` does not exist, for the reason I mentioned: offsetting a `NonNull` can result in null.

For a concrete example: What is the behavior of this safe code with your proposed semantics?

```rs
use std::{ptr::NonNull, mem::offset_of};
#[repr(C)]
struct Foo {
  x: u8,
  y: u8,
}

let ptr: NonNull<Foo> = NonNull::new(usize::MAX as *mut Foo);
let field: NonNull<u8> = ptr->y; // either this panics or returns the "wrong" pointer, or we have UB somewhere
assert_eq!(ptr.as_ptr().addr(), usize::MAX);
assert_eq!(offset_of!(Foo, y), 1); // Foo is repr(C)
assert_eq!(usize::MAX.wrapping_add(1), 0); // the expected address of (*ptr).y is 0, which is not a valid NonNull
assert_eq!(field.as_ptr().addr(), 0); // NonNull with addr 0?
```

> The reason for why I used `wrapping_add` is that the following code is UB when the `->` operation uses `add`.

Yes, offsetting outside of an allocation using field projections is what I am saying I think *should* be UB. Also, what "following code" are you referring to here? The next code snippet after this line of the file is the `set_flags_raw` example, which `unsafe`ly writes to the projected pointer anyway and does not discuss how making the projection safe reduces UB (though it could in some edge cases; I just don't think they are worth it).

> Maybe we can split it into two different mutually exclusive traits, but I think it would be better both for consistency and for ergonomic reasons to keep field projections safe.

If we want this operator to be able to be used generically, and not just for concrete pointer types, we could have an `UnsafeFoo` trait be a supertrait of the `Foo` trait, such that any `Foo` is also usable in a `where T: UnsafeFoo` context (perhaps with a blanket `impl<T: ?Sized + Foo> UnsafeFoo for T { ... }` so users don't have to implement `UnsafeFoo` for their `Foo` types).

Reply:
> > Note that `wrapping_add` is itself a safe function. So the implementation of `NonNull::project` only contains safe code, making UB impossible. Also, there is no need for a panic, since a projection can never result in null.
> 
> `NonNull::wrapping_add` does not exist, for the reason I mentioned: offsetting a `NonNull` can result in null.

Oh, totally overlooked that. Yeah that seems bad...

> For a concrete example: What is the behavior of this safe code with your proposed semantics?

Yeah that doesn't really work...

> > The reason for why I used `wrapping_add` is that the following code is UB when the `->` operation uses `add`.
> 
> Yes, offsetting outside of an allocation using field projections is what I am saying I think _should_ be UB. 

Agreed.

> Also, what "following code" are you referring to here?

Oops, I meant to add some code, but the point is moot now, since I didn't understand what you meant before.

> The next code snippet after this line of the file is the `set_flags_raw` example, which `unsafe`ly writes to the projected pointer anyway and does not discuss how making the projection safe reduces UB (though it could in some edge cases; I just don't think they are worth it).
> 
> > Maybe we can split it into two different mutually exclusive traits, but I think it would be better both for consistency and for ergonomic reasons to keep field projections safe.
> 
> If we want this operator to be able to be used generically, and not just for concrete pointer types, we could have an `UnsafeFoo` trait be a supertrait of the `Foo` trait, such that any `Foo` is also usable in a `where T: UnsafeFoo` context (perhaps with a blanket `impl<T: ?Sized + Foo> UnsafeFoo for T { ... }` so users don't have to implement `UnsafeFoo` for their `Foo` types).

Yes, we could have:

```rust
pub trait UnsafeProject<F>: Projectable
where
    F: Field<Base = Self::Inner>,
{
    type Output;

    unsafe fn project(self) -> Self::Output;
}

impl<P, F> UnsafeProject<F> for P
where
    P: Projectable,
    F: Field<Base = <P as Projectable>::Inner>,
    P: Project<F>,
{
    type Output = <P as Project<F>>::Output;

    unsafe fn project(self) -> Self::Output {
        <P as Project<F>>::project(self)
    }
}
```

But I don't know if it is possible for the compiler to detect which trait is implemented.

Reply:
The compiler could just always use `Project` when available, and fallback to `UnsafeProject` otherwise.

This is similar how the compiler is able to choose which of `Fn`/`FnMut`/`FnOnce` when desugaring call to an unboxed closure. 

But I'm not sure if the compiler also required a hierarchical relationship similar to the Fn traits to make this work.

```rust
pub trait UnsafeProject<F>: Projectable
where
    F: Field<Base = Self::Inner>,
{
    type Output;
    unsafe fn project_unchecked(self) -> Self::Output;
}

pub trait Project<F>: UnsafeProject<F>
where
    F: Field<Base = Self::Inner>,
{
    fn project(self) -> Self::Output;
}
```

Reply:
Makes sense. We also have to coordinate this with [possibly `ProjectRef`/`ProjectMove`](https://github.com/rust-lang/rfcs/pull/3735#discussion_r1871549746)... Seems like the combinatorial explosion is going to be bad...